riscv_isac/coverage.py

```diff
@@ -723,55 +746,78 @@ def compute_per_line(instr, cgf, xlen, addr_pairs,  sig_addrs):
-                            elif instr.instr_name in ["fmadd.s","fmsub.s","fnmadd.s","fnmsub.s"]:
-                                    val_key = fmt.extract_fields(32, rs1_val, str(1))
+                            elif instr.instr_name in ["fmadd.s","fmsub.s","fnmadd.s","fnmsub.s","fmadd.d","fmsub.d","fnmadd.d","fnmsub.d"]:
+                                    val_key = fmt.extract_fields(int(arch_state.flen), rs1_val, str(1))
                                     val_key+= " and "
-                                    val_key+= fmt.extract_fields(32, rs2_val, str(2))
+                                    val_key+= fmt.extract_fields(int(arch_state.flen), rs2_val, str(2))
                                     val_key+= " and "
-                                    val_key+= fmt.extract_fields(32, rs3_val, str(3))
+                                    val_key+= fmt.extract_fields(int(arch_state.flen), rs3_val, str(3))
                                     val_key+= " and "
-                                    val_key+= 'rm == '+ str(rm_val)
+                                    val_key+= 'rm_val == '+ str(rm_val)
                                     l=[0]
-                                    l[0] = val_key
+                                    l[0] = val_key + "  #nosat"
                                     val_key = l
-                                    if(val_key[0] in cgf[cov_labels]['val_comb']):
+                                    tempString = str(cgf[cov_labels]['val_comb'])
+                                    tempSubStr = str(val_key[0])
+                                    #if(val_key[0] in cgf[cov_labels]['val_comb']):
+                                    if(tempString.find(tempSubStr) != -1 ):
                                         if cgf[cov_labels]['val_comb'][val_key[0]] == 0:
                                             stats.ucovpt.append(str(val_key[0]))
                                         stats.covpt.append(str(val_key[0]))
                                         cgf[cov_labels]['val_comb'][val_key[0]] += 1
+                            elif instr.instr_name in ["fcvt.s.wu","fcvt.s.w","fcvt.s.l","fcvt.s.lu","fcvt.l.s","fcvt.lu.s","fmv.w.x","fmv.x.w","fcvt.d.s","fcvt.d.w","fcvt.d.wu","fcvt.s.d","fsd","fmv.d.x","fmv.x.d","fcvt.l.d","fcvt.lu.d"]:
+                                if instr.is_rvp and "rs1" in value:
+                                    op_width = 64 if instr.rs1_nregs == 2 else xlen
+                                    simd_val_unpack(value['val_comb'], op_width, "rs1", rs1_val, lcls)
+                                if instr.is_rvp and "rs2" in value:
+                                    op_width = 64 if instr.rs2_nregs == 2 else xlen
+                                    simd_val_unpack(value['val_comb'], op_width, "rs2", rs2_val, lcls)
+                                for coverpoints in value['val_comb']:
+                                    if cgf[cov_labels]['val_comb'][coverpoints] == 0:
```

Comment:
What is this specific block intended to do? It doesn't seem to evaluate any coverpoints for these instructions. No coverage metrics are being updated. Only the data propagation statistics are being updated to reveal unmet coverpoints?

Reply:
This block is specifically to handle few of the finstructions which are likely to fall inside the final else block. This is again to eliminate the eval, as it is bit ambiguous the way how we implemented on the else block.

Reply:
This entire block of code is wrong. On line 820, the coverpoints are being updated unconditionally? No checks to see if the coverpoints are actually satisfied. The `eval` based strategy is generic and tests whether all the fields relevant to the coverpoint have values which satisfy the boolean equation of the coverpoint.